File: 02/site/src/lib/methods/falsePosition.ts
/**
 * Method of False Position (Regula Falsi) — §2.4.
 * Like bisection, but uses the secant intercept p_k = a_k - f(a_k) · (a_k - b_k)/(f(a_k) - f(b_k))
 * instead of the midpoint.
 */
export interface FalsePosStep {
  k: number
  a: number
  b: number
  p: number
  fp: number
}

export interface FalsePosOptions {
  a: number
  b: number
  tol: number
  maxIter: number
}

export interface FalsePosResult {
  trace: FalsePosStep[]
  converged: boolean
  reason: 'tol' | 'exact' | 'maxiter' | 'invalid'
}

export function falsePosition(
  f: (x: number) => number,
  opts: FalsePosOptions,
): FalsePosResult {
  let { a, b } = opts
  const { tol, maxIter } = opts
  const trace: FalsePosStep[] = []

  let fa = f(a)
  let fb = f(b)
  if (!(fa * fb < 0)) {
    return { trace, converged: false, reason: 'invalid' }
  }

  let prevP = Number.NaN
  for (let k = 0; k <= maxIter; k++) {
    const p = a - (fa * (a - b)) / (fa - fb)
    const fp = f(p)
    trace.push({ k, a, b, p, fp })

    if (fp === 0) {
      return { trace, converged: true, reason: 'exact' }
    }
    if (k > 0 && Math.abs(p - prevP) < tol) {
      return { trace, converged: true, reason: 'tol' }
    }

    if (fa * fp < 0) {
      b = p
      fb = fp
    } else {
      a = p
      fa = fp
    }
    prevP = p
  }
  return { trace, converged: false, reason: 'maxiter' }
}

File: src/pages/practice/guides/matlab/false-position.m
function [c, k] = false_position(f, a, b, tol, max_iter)
% FALSE_POSITION  Regula falsi (bracketing secant intercept).
    if nargin < 4, tol = 1e-12; end
    if nargin < 5, max_iter = 200; end
    fa = f(a); fb = f(b); c = a;
    for k = 1:max_iter
        c = (a*fb - b*fa)/(fb - fa); fc = f(c);
        if abs(fc) < tol, return; end
        if fa*fc < 0, b = c; fb = fc; else, a = c; fa = fc; end
    end
end

% --- Demo ---
disp(false_position(@(x) x^2 - 2, 1.0, 2.0));  % sqrt(2)
